File: binary-search/binary-search.c
#include <stdio.h>
//the approach I use here is recursive 
//extra assumptions were made such as knowledge of first and last index
//this way the indexes markers do not have to be initialized at the beginnig of the function

int binarySearch(int arr[], int target, int l, int r){
    if (l <= r){ //base case
        int middle = (l+r)/2; //find the middle
        if (target == arr[middle]){ //self explanatory
            printf("Element was found at index %i.\n", middle);
            return 0;
        }
        else if(target < arr[middle]) //omit the right half by modifying the index markers
            return binarySearch(arr, target, l, middle - 2);
        else
            return binarySearch(arr, target, middle+1, r); //omit the left half 
    }
    
    printf("Element was not found. \n"); //first condition wasn't met meaning the left marker exceeded the right one which makes arr[midddle] = null
    return -1;
}

int main(void){ //test
    int arr[] = {0, 1, 3, 5, 6, 7, 10};
    binarySearch(arr, 7, 0, 6);
    binarySearch(arr, 2, 0, 6);

}

File: binary-search/binary-search.js
const values = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] //case 1 ODD NO. of VALUES
const values_2 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] //case 2 EVEN 


const binary = (target, arr) => {
    let lower = 0; //here we set initials markers for U&L bounds.
    let upper = arr.length - 1; 

    while (lower <= upper){
        let middle = Math.floor((upper+lower)/2); //truncate to get a middle marker
        if (target === arr[middle]){
            console.log("Element was found at index",middle);
            return 0; //self explanatory
        }
        else if (target < arr[middle]){ //since array is sorted target should be on left here
            upper = middle - 1; //set a new U marker to omit the right half 
        }
        else{
            lower = middle + 1; //set a new L marker to omit the left half
        }
    }


//If we reach this point it means that we kept omitting halves until we were down to one number and still we did not return the middle which at the last iteration was equal to this one number.
    console.log("NOT FOUND"); 
    return -1;
}


//TESTS
binary(2, values);
binary(11, values_2);
binary(1, values_2);

